internal/adapters/consul/http.go

```diff
@@ -281,42 +291,54 @@ func (f *flattenedRouteMap) flatten(route core.HTTPRoute) {
 					},
 				}}
 			}
+
+			// Add all matches for this rule to the list for this hostname
 			for _, match := range rule.Matches {
-				found = append(found, flattenedRoute{
+				matches = append(matches, hostnameMatch{
 					match:    match,
 					filters:  rule.Filters,
 					services: rule.Services,
 				})
 			}
 		}
-		f.flattenedRoutesByHostname[host] = found
+
+		f.matchesByHostname[host] = matches
 	}
 }
 
-func (f *flattenedRouteMap) constructRoutes(gateway core.ResolvedGateway) []core.HTTPRoute {
-	coreRoutes := []core.HTTPRoute{}
-	for hostname, routes := range f.flattenedRoutesByHostname {
-		sort.SliceStable(routes, func(i, j int) bool {
-			return compareHTTPRules(routes[i].match, routes[j].match)
-		})
-		rules := []core.HTTPRouteRule{}
-		for _, match := range routes {
-			rules = append(rules, core.HTTPRouteRule{
-				Matches:  []core.HTTPMatch{match.match},
-				Filters:  match.filters,
-				Services: match.services,
-			})
-		}
-		name := gateway.ID.Service + "-" + hostsKey(hostname)
-		coreRoutes = append(coreRoutes, core.HTTPRoute{
+// consolidateRoutes combines all rules into the shortest possible list of routes
+// with one route per hostname containing all rules for that hostname.
+func (f *routeConsolidator) consolidateRoutes(gateway core.ResolvedGateway) []core.HTTPRoute {
```

Comment:
Since we're already in the process of simplifying some of the naming conventions 😄 

```suggestion
func (f *routeConsolidator) consolidate(gateway core.ResolvedGateway) []core.HTTPRoute {
```